README.md
<img src="https://avatars1.githubusercontent.com/u/56512535?s=200&v=4" align="right" />

# Rhino.fi Javascript Trading API

A js client library for Rhino.fi - StarkWare orders

**Note:** This library is for Rhino.fi. A test version of the platform to use during integrations is connected to the Goerli test network and hosted at:
- UI: https://app.stg.rhino.fi
- API: https://api.stg.rhino.fi

## Contents

<!--ts-->
* [Rhino.fi Javascript Trading API](#rhinofi-javascript-trading-api)
   * [Contents](#contents)
   * [Setup](#setup)
      * [Pre Requisites](#pre-requisites)
      * [Instatiating](#instatiating)
         * [Using MetaMask or a local node](#using-metamask-or-a-local-node)
         * [Using a private key](#using-a-private-key)
         * [Configuration](#configuration)
            * [Parameters](#parameters)
      * [API Authentication](#api-authentication)
            * [Parameters](#parameters-1)
      * [Registering](#registering)
            * [Parameters](#parameters-2)
      * [Approving Tokens](#approving-tokens)
            * [Parameters](#parameters-3)
      * [Depositing tokens](#depositing-tokens)
            * [Parameters](#parameters-4)
      * [Placing an order](#placing-an-order)
            * [Parameters](#parameters-5)
      * [Getting Orders](#getting-orders)
            * [Parameters](#parameters-6)
      * [Cancelling Orders](#cancelling-orders)
            * [Parameters](#parameters-7)
      * [Withdrawing tokens](#withdrawing-tokens)
         * [Requesting a withdrawal](#requesting-a-withdrawal)
            * [Parameters](#parameters-8)
         * [Withdraw on chain](#withdraw-on-chain)
            * [Parameters](#parameters-9)
      * [Authenticated data endpoints](#authenticated-data-endpoints)
            * [Parameters](#parameters-10)
   * [More Examples](#more-examples)
      * [Gas Price](#gas-price)
      * [Custom order ID](#custom-order-id)
   * [Troubleshooting](#troubleshooting)
   * [Links](#links)
   * [Developing](#developing)
      * [Setup](#setup-1)
      * [Implementing a new feature and Testing](#implementing-a-new-feature-and-testing)
   * [License](#license)

<!-- Created by https://github.com/ekalinin/github-markdown-toc -->
<!-- Added by: adrian, at: Mon  6 Feb 11:54:24 GMT 2023 -->

<!--te-->

## Setup

### Pre Requisites

  - An Ethereum wallet
  - A web3 provider with your account or a private key
    * Such as MetaMask, keystore file, hardware wallet or raw private key

### Instatiating

#### Using MetaMask or a local node

```javascript
// In case of MetaMask make sure you call ethereum.enable() before using it
const RhinofiClientFactory = require('@rhino.fi/client-js')
const rhinofi = await RhinofiClientFactory()
```

#### Using a private key

```javascript
const HDWalletProvider = require("@truffle/hdwallet-provider");
const Web3 = require("Web3")

const privateKey = '8F085...' // Account's private key
const rpcUrl = 'https://mainnet.infura.io/v3/9e28b...'

const provider = new HDWalletProvider(privateKey, rpcUrl)
const web3 = new Web3(provider)

rhinofi = await RhinofiClientFactory(web3)
```

For more, see [examples](/examples) and their [README](/examples/README.md).

#### Configuration

It's possible to overwrite values on the configuration on a per instance basis.

The [default configuration](./src/config.js) can be overwritten with an optional
parameter `userConf` when calling the RhinofiClientFactory function.

##### Parameters

- `api` **[string](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/String)**? (default `https://api.stg.rhino.fi`) API endpoint you are connecting to Staging (ropsten): https://api.stg.rhino.fi, Production (mainnet): https://api.rhino.fi)
- `gasApi` **[string](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/String)**? (default `https://ethgasstation.info`)
- `defaultGasLimit` **[number](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Number)**? (default `200000`)
- `defaultGasPrice` **[number](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Number)**? (default `50000000000`)
- `defaultStarkExpiry` **[number](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Number)**? (default `720`) Expiration time for transfers and orders in hours
- `defaultNonceAge` **[number](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Number)**? (default `43200`) Nonce age in seconds
- `defaultProvider` **[string](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/String)**? (default `http://localhost:8545`) In case no web3 provider is provided we will try connecting to this default
- `autoLoadUserConf` **[boolean](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Boolean)**? - Enables integrators to select if they want to call `getUserConfig` upon initialization
- `autoLoadExchangeConf` **[boolean](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Boolean)**? - Enables integrators to select if they want to call `getConfig` upon initialization

For instance:

```javascript
  rhinofi = await RhinofiClientFactory(web3, {
    api: 'https://your-custom-api-address',
    gasStationApiKey: 'a1b2c3...'
  })
```

The configuration is also merged with the configuration provided by the exchange
on the HTTP endpoint `/v1/trading/r/getConf` which at the moment looks similar
to this:

```json
{
   "RhinofiClientFactory":{
      "defaultFeeRate":0.002,
      "deversifiAddress":"0xaf8ae6955d07776ab690e565ba6fbc79b8de3a5d",
      "starkExContractAddress":"0x5d22045DAcEAB03B158031eCB7D9d06Fad24609b",
      "withdrawalBalanceReaderContractAddress":"0x650ca2dca7e2e2c8be3bb84e0a39dd77891d4d1e",
      "exchangeSymbols":[
         "ETH:USDT",
         "MKR:ETH",
         "MKR:USDT"
      ],
      "tempStarkVaultId":1,
      "minDepositUSDT":1
   },
   "tokenRegistry":{
      "ETH":{
         "decimals":18,
         "quantization":10000000000,
         "minOrderSize":0.05,
         "starkTokenId":"0xb333e3142fe16b78628f19bb15afddaef437e72d6d7f5c6c20c6801a27fba6"
      },
      "MKR":{
         "decimals":18,
         "quantization":10000000000,
         "minOrderSize":0.025,
         "tokenAddressPerChain": {
            "ETHEREUM": "0x9f8f72aa9304c8b593d555f12ef6589cc3a579a2"
         },
         "starkTokenId":"0x1a4af39d27ce2e3445ed084809e5bc36d03918df04b7e2b6ee3c769a9892600"
      }
   }
}
```

The complete compiled configuration is accessible through `rhinofi.config`, for instance:

```javascript
const rhinofi = await RhinofiClientFactory()

const config = rhinofi.config
```

### API Authentication

Authentication to make all authenticated requests is done by signing a `nonce` using an
Ethereum private key. Signing is handled by the Rhino.fi client
library if the account is available and unlocked or if the web3 provider supports it.
Otherwise the message and signature need to be prepared separately.

##### Parameters
- `nonce` **[string](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/String)** Nonce which is used to provide the time until which this nonce is valid. It is presented as seconds since epoch.
- `signature` **[string](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/String)** The signature obtained by signing the nonce with your private ethereum key.

```javascript
const nonce = (Date.now() / 1000).toString()
const signature = await rhinofi.sign(nonce)
```

### Registering
This method is used to register a stark public key that corresponds to an Ethereum public address or a trading key.


##### Parameters
- `starkPublicKey` **[Object](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/Object)**
    - `x` **[string](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/String)** - First 32 bits of stark public key
- `nonce` **[string](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/String)** Nonce which is used to provide the time until which this nonce is valid. It is presented as seconds since epoch.
- `signature` **[string](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/String)** The signature obtained by signing the nonce with your private ethereum key.
- `contractWalletAddress` **[string](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/String)**? Address of the deployed contract wallet (only for contract wallet integrations)

Returns **[Promise](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/Promise)&lt;[UserConfigResponse](https://docs.rhino.fi/docs#postV1TradingRGetuserconf)>**

### Approving Tokens

When depositing an ERC20 Ethereum-based token for the first time from a specific account,
you are required to approve it to interact with the smart contracts, this is not required for ETH.

##### Parameters
- `token` **[string](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/String)** Token symbol that's available in `rhinofi.config.tokenRegistry`

Returns **[Promise](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/Promise)&lt;[PromiEvent](https://web3js.readthedocs.io/en/v1.2.11/callbacks-promises-events.html#promievent)>**

```javascript
const token = 'ETH'
await rhinofi.contract.approve(token)
```

This step does not need to be repeated again, and subsequently you are required
only to call the deposit function.

### Depositing tokens
This method is used to deposit the tokens to the smart contract and submit a signed notification of a new deposit made to the API.

##### Parameters
- `token` **[string](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/String)** Token symbol available in `rhinofi.config.tokenRegistry` to be deposited
- `amount` **[number](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Number)** || **[string](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/String)** Amount of tokens to be deposited
- `starkPrivateKey` **[string](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/String)** Trading key
- `nonce` **[string](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/String)**? Nonce which is used to provide the time until which this nonce is valid. It is presented as seconds since epoch.
- `signature` **[string](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/String)**? The signature obtained by signing the nonce with your private ethereum key.

Returns **[Promise](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/Promise)&lt;{...[PromiEvent](https://web3js.readthedocs.io/en/v1.2.11/callbacks-promises-events.html#promievent), ...[DepositResponse](https://docs.rhino.fi/docs#postV1TradingWDeposit)}>**

```javascript
const token = 'ETH'
const amount = 100

const deposit = await rhinofi.deposit(token, amount, tradingKey)
```

### Placing an order

This authenticated endpoint is used to place an order.

##### Parameters
- `symbol` **[string](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/String)** Pair which you wish to trade
- `amount` **[number](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Number)** || **[string](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/String)** Order amount specified in the first currency in the symbol (i.e. ZRXETH). For a sell, amount is negative. Amount accepts either maximum 8 d.p, or as many decimals as are available on the relevant token's smart contract if it is fewer than 8.
- `price` **[number](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Number)** || **[string](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/String)** Order price  specified in the second currency in the symbol (i.e. ZRXETH). Prices should be specified to 5 s.f. maximum.
- `nonce` **[string](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/String)**? Nonce which is used to provide the time until which this nonce is valid. It is presented as seconds since epoch.
- `signature` **[string](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/String)**? The signature obtained by signing the nonce with your private ethereum key.
- `starkPrivateKey` **[string](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/String)**? Trading key (for keystore etc.)
- `ledgerPath` **[string](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/String)**? Ledger derivation path if using ledger
- `isPostOnly` **[boolean](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/Boolean)?** Flag to indicate if the order is post-only.
- `isHidden` **[boolean](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/Boolean)?** Flag to indicate if the order is hidden.
- `validFor` **[number](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Number)**? Validation time in hours
- `feeRate` **[number](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Number)**? Fee rate if known
- `cid` **[string](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/String)**? Optional custom order ID that could be set when placing order and used later to retrieve order. This ID is unique per user (user A and B can each have an order with `cid = AAA`, but the same user cannot have two orders with the same `cid`). [See example](#custom-order-id)
- `gid` **[string](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/String)**?
- `partnerId` **[string](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/String)**?
- `ethAddress` **[string](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/String)**?
- `type` **[string](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/String)**? Order type (`EXCHANGE LIMIT`, `EXCHANGE MARKET`)
- `protocol` **[string](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/String)**? (default `stark`)


Returns **[Promise](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/Promise)&lt;[SubmitOrderResponse](https://docs.rhino.fi/docs#postV1TradingWSubmitorder)>**

```javascript
const symbol = 'NEC:ETH'
const amount = -15
const price = 0.0025

const orderId = await rhinofi.submitOrder(symbol, amount, price)
```

### Getting Orders
This method allows you to get a specific order by `orderId` or `cid`.

##### Parameters
- `orderId` **[string](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/String)**? ID of the order
- `nonce` **[string](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/String)**? Nonce which is used to provide the time until which this nonce is valid. It is presented as seconds since epoch.
- `signature` **[string](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/String)**? The signature obtained by signing the nonce with your private ethereum key.
- `cid` **[string](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/String)**? If order was placed with custom order ID (`cid`) property set, it can be canceled using same `cid`.

Returns **[Promise](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/Promise)&lt;[CancelOrderResponse](https://docs.rhino.fi/docs#postV1TradingWCancelorder)>**

```javascript
const orderID = '123'
const customID = 'cid-123'

const order = await rhinofi.getOrder({ orderId: orderID })
// or
const order = await rhinofi.getOrder({ cid: customID })
```


### Cancelling Orders
This method allows you to cancel a specific order by `orderId` or `cid`.

##### Parameters
- `orderId` **[string](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/String)** ID of the order
- `nonce` **[string](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/String)**? Nonce which is used to provide the time until which this nonce is valid. It is presented as seconds since epoch.
- `signature` **[string](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/String)**? The signature obtained by signing the nonce with your private ethereum key.
- `cid` **[string](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/String)**? If order was placed with custom order ID (`cid`) property set, it can be canceled using same `cid`.

Returns **[Promise](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/Promise)&lt;[CancelOrderResponse](https://docs.rhino.fi/docs#postV1TradingWCancelorder)>**

```javascript
const orderID = '123'
const customID = 'cid-123'

const response = await rhinofi.cancelOrder({ orderId: orderID })
// or
const response = await rhinofi.cancelOrder({ cid: customID })
```

### Withdrawing tokens
#### Requesting a withdrawal
This method submits a request for a new withdrawal.
##### Parameters
- `recipientEthAddress` **[string](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/String)** Trading key
- `token` **[string](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/String)** Token symbol available in `rhinofi.config.tokenRegistry` to be withdrawn
- `amount` **[number](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Number)** || **[string](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/String)** Amount of tokens to be withdrawn

Returns **[Promise](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/Promise)&lt;[WithdrawResponse](https://docs.rhino.fi/docs#postV1TradingWWithdraw)>**

```javascript
const token = 'ETH'
const amount = 100
const withdrawal = await await rhinofi.transferAndWithdraw({
  recipientEthAddress: address,
  token,
  amount
})
```

#### Withdraw on chain
This method calls the contract and withdraws the tokens to your wallet
##### Parameters
- `token` **[string](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/String)** Token symbol available in `rhinofi.config.tokenRegistry` to be withdrawn

Returns **[Promise](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/Promise)&lt;{ transactionHash: **[string](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/String)** }>**

```javascript
const token = 'ETH'
const txHash = await rhinofi.withdrawOnchain(token, recipientEthAddress)
```

### Authenticated data endpoints

If you already have an unlocked wallet available to web3 to use for signing,
you can simply get data from the API as follows:

Note: You should reuse the `nonce` and `signature` and pass them to these methods while they are valid to avoid unnecessary signing

##### Parameters
- `nonce` **[string](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/String)**? Nonce which is used to provide the time until which this nonce is valid. It is presented as seconds since epoch.
- `signature` **[string](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/String)**? The signature obtained by signing the nonce with your private ethereum key.

```javascript
// Get all open orders
const openOrders = await rhinofi.getOrders()

// Get all historical orders
const historicalOrders = await rhinofi.getOrdersHist()

// Get specific order
const id = "123"
const order = await rhinofi.getOrder(id)

// Get exchange balances
const balance = await rhinofi.getBalance()

// Get deposits
const deposits = await rhinofi.getDeposits()

// Get withdrawals
const withdrawals = await rhinofi.getWithdrawals()

// Get user config
const userConfig = await rhinofi.getUserConfig()
```

## More Examples

Aside from these examples, there are complete examples in the [examples folder](./examples)

### Gas Price

You can setup a default custom gas price by setting up the 'defaultGasPrice' property
```javascript
const rhinofi = await RhinofiClientFactory()

rhinofi.set('defaultGasPrice', web3.utils.toWei('2', 'gwei'))

```
RhinofiClientFactory calls https://ethgasstation.info API to get the current gas prices and calculate a safe gas price for Ethereum transactions. Access to the ETH Gas Station API is free, but rate limited if you are not using an API key. If a ETH Gas Station API key is not provided then a recommended gas price is used which is available in `rhinofi.recommendedGasPrices`.

To configure your api key with rhinofi client please pass this as a `userConf` parameter when initialising RhinofiClientFactory:

```
javascript
  rhinofi = await RhinofiClientFactory(web3, {
    gasStationApiKey: 'a1b2c3...'
  })
```
or by setting the 'gasStationApiKey' property:

```javascript

rhinofi.set('gasStationApiKey', 'a1b2c3...')

```
### Custom order ID
Property `cid` can be used to give order custom identificator for further tracking.

```js
const symbol = 'ETH:USDT'
const amount = -1.42
const price = 3000

const customOrderID = `short-` + Math.random().toString(36).substring(7)

await rhinofi.submitOrder({
  symbol,
  amount,
  price,
  cid: customOrderID,
})

// ...
// Later we can use `cid` to get order
const order = await rhinofi.getOrder({cid: customOrderID})

// or cancel it
await rhinofi.cancelOrder({cid: customOrderID})
```

## Troubleshooting

A list of error codes returned by the API and reasons are available [here](./src/lib/dvf/errorReasons.js#L1).
Some more detailed explanations can also be found in the [API Documentation](https://docs.rhino.fi).

If you have suggestions to improve this guide or any of the available
documentation, please raise an issue on Github, or email [feedback@rhino.fi](mailto:feedback@rhino.fi).

## Links

 - [API documentation](https://docs.rhino.fi)

## Developing

### Setup

1. [install nix](https://nixos.org/download.html)
2. enter nix shell:
  ```sh
  $ nix-shell
  ```

3. install js deps
  ```
  $ yarn
  ```

### Implementing a new feature and Testing

TODO

## License

MIT


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).


## 5.0.0 - 2023-01-07
### Changed
- rename from dvf-client-js to @rhino.fi/client-js
- rename entry point from src/dvf to src/index
- remove references to dvf from README and examples

## 4.2.0 - 2023-01-06
### Changed
- Add new methods
- Use yarn berry
- Update\cleanup README.md
- Update node deps
- Add shell.nix etc

## 4.1.0 - 2022-06-30
### Changed
- Support fee on transferAndWithdraw

## 4.0.0 - 2022-06-15
### Changed
- StarkEx v3 compatibility removed along with onchain registration functions

## 3.1.9 - 2021-05-16
### Changed
- EIP712 signature helper for ledger devices

## 3.1.8 - 2021-04-12
### Changed
- Add slippage reduction for funding stark orders

## 3.1.7 - 2021-01-15
### Changed
- Store input "web3" as both ETHEREUM and DEFAULT web3 provider when not specified

## 3.1.6 - 2021-01-12
### Changed
- Update bridge contract addresses conf location

## 3.1.5 - 2021-01-07
### Changed
- Update `withdrawOnChain` example for StarkEx V4

## 3.1.3 - 2021-11-15
### Changed
- Add getAmmFundingOrder method to track AMM operations

## 3.1.2 - 2021-11-10
### Changed
- Add poolTvlHistory endpoint

## 3.1.1 - 2021-11-10
### Changed
- When user denies register TX, respond with more friendly message

## 3.1.0 - 2021-11-04
### Changed
- Assuming default web3 provider is an Ethereum mainnet or testnet
- `useAuthHeader` authentication mode for all examples

## 3.0.3 - 2021-10-22
### Changed
- Add manual xDVF support

## 3.0.2 - 2021-09-24
### Changed
- Fixed issue preventing native tokens to be deposited from sidechains

## 3.0.1 - 2021-10-06
### Changed
- Updated Ledger data provision for Goerli test tokens (USDT, WBTC, DVF, xDVF)

## 3.0.0 - 2021-09-21
### Changed
- Using RPC_URL instead of INFURA_PROJECT_ID in config for more flexibility

## 2.8.3 - 2021-09-16
### Changed
- Update feeRate check on order payload creation to replace feeRate only if not present and preserve falsy values like 0 to allow staking

## 2.8.2 - 2021-08-02
### Changed
- Bridged deposit to call deposit validation endpoint for sanity checks

## 2.8.1 - 2021-07-14
### Changed
- Fixed issues causing Ledger transactions to require reload after error

## 2.8.0 - 2021-07-09
### Changed
- Adding DVF token metadata for Ledger compatibility

## 2.7.0 - 2021-06-30
### Changed
- Support of multiple web3 providers passed in the constructor

## 2.6.0 - 2021-06-17
### Changed
- Add `bridgedDeposit` and `bridgedWithdraw` methods

## 2.5.2 - 2021-06-11
### Changed
- Use GET method when getting gas price from API in `getGasPrice`

## 2.5.1 - 2021-06-09
### Changed
- Add optional `memo` to transfer API

## 2.5.0 - 2021-06-7
### Changed
- Adding optional `permitParams` argument to `depositV2` and `registerAndDeposit` (will try to use signature-based eip-2612 permit instead of `approve` if passed)

## 2.4.0 - 2021-06-4
### Changed
- Accomodating changes in configuration tokenRegistry structure (`tokenAddressPerChain`)
- `dvf.approve` to accept an extra argument `chain` (defaults to `'ETHEREUM'`)
- accept transaction hash callback option when doing token approval, deposit and register + deposit

## 2.3.0 - 2021-05-25
### Changed
- `dvf.depositV2` and `dvf.registerAndDeposit` to accept optional `web3Options` parameter for internal use (better gas limit estimates)

## 2.2.0 - 2021-05-10
### Changed
- `dvf.getFeeRate` to accept `symbol` and `feature` parameters

## 2.1.4 - 2021-04-26
### Changed
- Public user permissioning added to the client

## 2.1.3 - 2021-16-04
### Added
- `dvf.getMinMaxOrderSize` to get order size limitations by `symbol`

## 2.1.2 - 2021-04-16
### Changed
- `dvf.register` and `dvf.registerAndDeposit` to accept an optional `meta` parameter for interal use

## 2.1.1 - 2021-04-09
### Changed
- Exposing `dvf.getAuthenticated` for arbitrary calls to GET endpoints
- Allowing custom headers in `dvf.getAuthenticated`

## 2.1.0 - 2021-03-31
### Added
- `dvf.getRegistrationStatuses` endpoint for checking both Deversifi and on-chain registration statuses - see `examples/29.getRegistrationStatuses.js`

## 2.0.1 - 2021-03-30
### Fixed
- Schema validated using `validateWithJoi` to be `required()` by default

## 2.0.0 - 2021-03-30
### Changed
- `dvf.transfer` method signature changes : `starkPrivateKey` to be set via configuration instead of input argument - see `examples/27.transfer.js`
- Updated examples

## 1.0.6 - 2020-07-08
### Fixed
- Handle errors in getConfig during initialisation

## 1.0.5 - 2020-07-02
### Added
- Use dvf pub api to get range of gas prices

## 1.0.4 - 2020-07-01
### Added
- gas Station API optional
- Use recommended gas price from config

## 1.0.3 - 2020-06-26
### Added
- Use Deversifi Symbol format in config file

## 1.0.2 - 2020-06-25
### Added
- Use API Key for eth gas station api call

## 1.0.1 - 2020-06-24
### Fixed
- web3 provider calls for blocknumber


src/config.js
module.exports = {
  // for production use 'https://api.rhino.fi',
  api: 'https://api.stg.rhino.fi',
  gasApi: 'https://ethgasstation.info',

  // default transaction arguments
  defaultGasLimit: 300000,
  defaultGasPrice: 50000000000,

  // default stark related constants
  // default expiration time for transfers and orders in hours
  defaultStarkExpiry: 4320,
  // default nonce age in seconds
  defaultNonceAge: 43200,
  // in case no provider is provided we will try connecting to the this default

  // address
  defaultProvider: 'http://localhost:8545',

  // default account to select in case no account is provided by the userConfig
  // parameter
  account: 0,
  // selects account from web3 provider based on config.account upon
  // initialization
  autoSelectAccount: true,

  // Ledger Stark Path Constants
  purpose: 2645,
  plugin: 579218131,
  application: 1393043894,
  accountIndex: 0,


  // enables integrators to select if they want to fetch user config upon initialization
  autoLoadUserConf: true,
  autoLoadExchangeConf: true,

  // enables integrations exposing starkProvider
  starkProvider: null
}
